File: libraries/python/getpatter/tts/soniox.py
"""Soniox TTS for Patter pipeline mode."""

from __future__ import annotations

import os
from typing import ClassVar, Optional

from getpatter.providers.soniox_tts import SonioxTTS as _SonioxTTS

__all__ = ["TTS"]


def _resolve_api_key(api_key: str | None) -> str:
    # Shared credential with Soniox STT — same SONIOX_API_KEY env var.
    key = api_key or os.environ.get("SONIOX_API_KEY")
    if not key:
        raise ValueError(
            "Soniox TTS requires an api_key. Pass api_key='...' or "
            "set SONIOX_API_KEY in the environment."
        )
    return key


class TTS(_SonioxTTS):
    """Soniox HTTP TTS (``tts-rt-v1``, default voice ``Adrian``).

    Shares the ``SONIOX_API_KEY`` credential with the Soniox STT adapter.

    Example::

        from getpatter.tts import soniox

        tts = soniox.TTS()                          # reads SONIOX_API_KEY
        tts = soniox.TTS(api_key="...", voice="Maya", language="en")

    Telephony optimization
    ----------------------
    Use :meth:`for_twilio` or :meth:`for_telnyx` on phone calls. Both emit
    mu-law @ 8 kHz natively (Soniox supports G.711 directly), so the pipeline
    skips resampling and PCM -> mu-law encoding entirely (bit-clean
    passthrough).
    """

    provider_key: ClassVar[str] = "soniox_tts"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        model: str = "tts-rt-v1",
        voice: str = "Adrian",
        language: str = "en",
        audio_format: str = "pcm_s16le",
        sample_rate: int = 16000,
        bitrate: Optional[int] = None,
        speed: Optional[float] = None,
    ) -> None:
        super().__init__(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            audio_format=audio_format,
            sample_rate=sample_rate,
            bitrate=bitrate,
            speed=speed,
        )

    @classmethod
    def for_twilio(
        cls,
        api_key: str | None = None,
        *,
        model: str = "tts-rt-v1",
        voice: str = "Adrian",
        language: str = "en",
        speed: Optional[float] = None,
    ) -> "TTS":
        """Pipeline TTS pre-configured for Twilio Media Streams.

        Emits mu-law @ 8 kHz natively — Twilio's wire codec — so the pipeline
        passes the bytes straight through (no resample, no PCM -> mu-law
        encode). Falls back to ``SONIOX_API_KEY`` from the env when ``api_key``
        is omitted.
        """
        return cls(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            audio_format="pcm_mulaw",
            sample_rate=8000,
            speed=speed,
        )

    @classmethod
    def for_telnyx(
        cls,
        api_key: str | None = None,
        *,
        model: str = "tts-rt-v1",
        voice: str = "Adrian",
        language: str = "en",
        speed: Optional[float] = None,
    ) -> "TTS":
        """Pipeline TTS pre-configured for Telnyx bidirectional media.

        Emits mu-law @ 8 kHz natively — the SDK pins the Telnyx wire to
        PCMU/mu-law @ 8 kHz — so audio flows end-to-end with zero resampling or
        transcoding. Falls back to ``SONIOX_API_KEY`` from the env when
        ``api_key`` is omitted.
        """
        return cls(
            api_key=_resolve_api_key(api_key),
            model=model,
            voice=voice,
            language=language,
            audio_format="pcm_mulaw",
            sample_rate=8000,
            speed=speed,
        )

File: libraries/typescript/src/tts/soniox.ts
/** Soniox TTS for Patter pipeline mode. */
import {
  SonioxTTS as _SonioxTTS,
  SonioxTTSAudioFormat,
  type SonioxTTSModel,
  type SonioxTTSSampleRate,
} from '../providers/soniox-tts';

/** Constructor options for the Soniox `TTS` adapter. */
export interface SonioxTTSOptions {
  /** API key. Falls back to SONIOX_API_KEY env var when omitted (shared with Soniox STT). */
  apiKey?: string;
  model?: SonioxTTSModel | string;
  voice?: string;
  language?: string;
  /** Audio encoding. Default `pcm_s16le`; `pcm_mulaw` @ 8 kHz = carrier-native. */
  audioFormat?: SonioxTTSAudioFormat | string;
  sampleRate?: SonioxTTSSampleRate | number;
  bitrate?: number;
  speed?: number;
  baseUrl?: string;
}

/** Options for the carrier-specific factories — same as the constructor minus `sampleRate`/`audioFormat`. */
export type SonioxCarrierOptions = Omit<
  SonioxTTSOptions,
  'sampleRate' | 'audioFormat'
>;

function resolveApiKey(apiKey: string | undefined): string {
  // Shared credential with Soniox STT — same SONIOX_API_KEY env var.
  const key = apiKey ?? process.env.SONIOX_API_KEY;
  if (!key) {
    throw new Error(
      "Soniox TTS requires an apiKey. Pass { apiKey: '...' } or " +
        'set SONIOX_API_KEY in the environment.',
    );
  }
  return key;
}

/**
 * Soniox TTS (`tts-rt-v1`, default voice `Adrian`).
 *
 * Shares the `SONIOX_API_KEY` credential with the Soniox STT adapter.
 *
 * @example
 * ```ts
 * import * as soniox from "getpatter/tts/soniox";
 * const tts = new soniox.TTS();                          // reads SONIOX_API_KEY
 * const tts = new soniox.TTS({ apiKey: "...", voice: "Maya", language: "en" });
 * ```
 *
 * **Telephony** — use {@link TTS.forTwilio} or {@link TTS.forTelnyx} on phone
 * calls. Both emit μ-law @ 8 kHz natively (Soniox supports G.711 directly), so
 * the pipeline skips resampling and PCM → μ-law encoding entirely (bit-clean
 * passthrough).
 */
export class TTS extends _SonioxTTS {
  static readonly providerKey = 'soniox_tts';
  constructor(opts: SonioxTTSOptions = {}) {
    const key = resolveApiKey(opts.apiKey);
    const { apiKey: _ignored, ...rest } = opts;
    void _ignored;
    super(key, rest);
  }

  /** Pipeline TTS pre-configured for Twilio Media Streams (μ-law @ 8 kHz native — carrier-wire passthrough, no resample/encode). */
  static override forTwilio(opts?: SonioxCarrierOptions): TTS;
  // Parent-compatible overload — accepts the legacy positional form too.
  static override forTwilio(apiKey: string, options?: SonioxCarrierOptions): TTS;
  static override forTwilio(
    arg1?: string | SonioxCarrierOptions,
    arg2?: SonioxCarrierOptions,
  ): TTS {
    const opts: SonioxCarrierOptions =
      typeof arg1 === 'string' ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({
      ...opts,
      audioFormat: SonioxTTSAudioFormat.PCM_MULAW,
      sampleRate: 8000,
    });
  }

  /** Pipeline TTS pre-configured for Telnyx (μ-law @ 8 kHz native — PCMU wire passthrough). */
  static override forTelnyx(opts?: SonioxCarrierOptions): TTS;
  static override forTelnyx(apiKey: string, options?: SonioxCarrierOptions): TTS;
  static override forTelnyx(
    arg1?: string | SonioxCarrierOptions,
    arg2?: SonioxCarrierOptions,
  ): TTS {
    const opts: SonioxCarrierOptions =
      typeof arg1 === 'string' ? { apiKey: arg1, ...(arg2 ?? {}) } : (arg1 ?? {});
    return new TTS({
      ...opts,
      audioFormat: SonioxTTSAudioFormat.PCM_MULAW,
      sampleRate: 8000,
    });
  }
}
